docs(concepts): Add documentation for Execution Listeners

ce-dmelnych opened the pull request:
## Description

Relates to: https://github.com/camunda/camunda/issues/16219

This PR adds documentation for Execution Listeners (EL) in Camunda 8, providing developers with a guide on how to implement and utilize ELs effectively in their workflows. 

It includes the following sections:
- **Introduction**: Overview, and use cases of ELs.
- **Getting Started**: Steps to enable ELs in BPMN models.
- **Limitations**: ELs limitations.
- **Usage**: How to define Job Workers to process EL jobs.
- **Variables**: Handling variables within ELs, including scope and dynamic variable handling.
- **References**: Links to additional documentation.

## When should this change go live?

<!-- PRs merged go to stage.docs.camunda.io first and must be manually released to docs.camunda.io. -->

- [ ] This is a bug fix, security concern, or something that needs **urgent release support**.
- [ ] This is already available but undocumented and should be released within a week.
- [ ] This on a **specific schedule** and the assignee will coordinate a release with the DevEx team. (apply `hold` label or convert to draft PR)
- [x] This is part of a scheduled alpha or minor. (apply alpha or minor label)
- [ ] There is **no urgency** with this change and can be released at any time.

## PR Checklist

<!-- Camunda maintains 18 months of versions. Backporting your change to multiple versions is common. -->

- [ ] My changes are for an **already released minor** and are in `/versioned_docs` directory.
- [x] My changes are for the **next minor** and are in `/docs` directory (aka `/next/`).

<!-- UNCOMMENT THIS SECTION IF APPLICABLE. Adding or removing pages requires extra steps.
- [ ] I included my new page in the sidebar file(s).
- [ ] I added a redirect for a renamed or deleted page to the .htaccess file.
-->

<!-- All changes require either an Engineering review or technical writer review. **Many require both!** -->

- [x] My changes require an [Engineering review](https://github.com/camunda/camunda-docs/blob/main/howtos/documentation-guidelines.md#review-process), and I've assigned the Engineering DRI or delegate.
- [x] My changes require a [technical writer review](https://github.com/camunda/camunda-docs/blob/main/howtos/documentation-guidelines.md#review-process), and I've assigned @camunda/tech-writers as a reviewer.

<!-- UNCOMMENT THIS SECTION IF APPLICABLE. Changes to **docs infra**, including updates to workflows and adding new npm packages, must be first discussed via issue or #ask-c8-documentation and linked for context.
- [ ] My changes require a [docs infrastructure review](https://github.com/camunda/camunda-docs/blob/main/howtos/documentation-guidelines.md#review-process). -->

ce-dmelnych commented:
> @ce-dmelnych Is this ready for review or not quite yet?

Hi @christinaausley ,

I have made some corrections, and the PR is now ready for review. However, it also requires an engineering review, so there might be some additional changes after that.

@saig0, could you please assist with the engineering review?

Thanks!

christinaausley commented:
@ce-dmelnych Cool 👍 I'll have a look after @saig0.

saig0 commented on docs/components/concepts/execution-listeners.md:
````diff
@@ -0,0 +1,352 @@
+    classDef execution-listeners fill:gray
+```
+
+<details>
+  <summary><strong>Element lifecycle steps</strong></summary>
+  <p>When an element with ELs is processed in Zeebe, the following steps outline its lifecycle, highlighting where <code>start</code> and <code>end</code> ELs are executed:</p>
+  <h4>Element Activation:</h4>
+  <p>The process instance reaches the element and begins its activation. This initial phase prepares the element for execution and typically includes:</p>
+  <ul>
+    <li><em>Apply Input Mappings</em> (if applicable): Input variables are mapped to the element's local scope.</li>
+  </ul>
+  <h4>Start Execution Listeners:</h4>
+  <p>If <code>start</code> ELs are defined, they are executed. These listeners can perform setup tasks and have access to input variables (if any defined).</p>
+  <h4>Finalize Activation:</h4>
+  <p>The element completes any additional actions required for activation. This usually includes:</p>
+  <ul>
+    <li><em>Subscribe to Events</em> (if applicable): The element subscribes to any events it needs to listen for during its execution.</li>
+    <li><em>Element Activated</em>: The element is fully activated and ready to perform its primary function.</li>
+  </ul>
+  <h4>Element Execution:</h4>
+  <p>The core activity of the element is carried out. This step represents the main purpose of the BPMN element (e.g., a task being performed).</p>
+  <h4>Element Completion:</h4>
+  <p>Once the primary activity is finished, the element begins its completion phase, which includes:</p>
+  <ul>
+    <li><em>Apply Output Mappings</em> (if applicable): Output variables are mapped from the element's local scope to the global scope.</li>
+    <li><em>Unsubscribe from Events</em> (if applicable): The element unsubscribes from any events it was listening for.</li>
+  </ul>
+  <h4>End Execution Listeners:</h4>
+  <p>During the element completion stage, if <code>end</code> ELs are defined, they are executed after applying output mappings (if any defined). These listeners can perform cleanup tasks and have access to output variables.</p>
+  <h4>Finalize Completion:</h4>
+  <p>The element executes any additional actions required for element completion and execution transits to the next element.</p>
+  <p>By placing custom logic at strategic points within the element lifecycle, you can enhance process control, improve data handling, and integrate seamlessly with external systems.</p>
+</details>
````
:wrench: This part is relevant to use EL correctly. We should not hide it.  

:wrench: We should make this part less technical and easier to understand for users.

saig0 commented on docs/components/concepts/execution-listeners.md:
```diff
@@ -0,0 +1,352 @@
+  <p>By placing custom logic at strategic points within the element lifecycle, you can enhance process control, improve data handling, and integrate seamlessly with external systems.</p>
+</details>
+
+## FAQ
+
+### Common questions
+
+- **Can execution listeners be used with all BPMN elements?**
+  - No, some elements have limitations due to their processing nature. See the table of [supported elements](#supported-elements).
+
+### Troubleshooting
+
+- **Execution Listener not triggering**
+  - Verify the configuration and ensure the element supports the specified listener type.
```
:wrench: No relevant information.

saig0 commented on docs/components/concepts/execution-listeners.md:
```diff
@@ -0,0 +1,352 @@
+By understanding these workarounds, you can effectively design workflows to accommodate the limitations of ELs
+and maintain the flexibility and control of your process executions.
+
+## Usage
+
+### Job Workers
+
+To process EL jobs, a Job Worker should be defined. Job Workers handle specific tasks by subscribing to jobs of a
+particular type and executing the associated business logic. They allow for flexible and reusable logic in workflows.
+For more detailed information on how to create and manage Job Workers,
+refer to the [Camunda Job Workers documentation](/components/concepts/job-workers.md) & [Job Worker example](apis-tools/java-client-examples/job-worker-open.md).
+
+## Variables
+
+### Scope
+
+Variables set during the completion of an EL job are local to the element where the EL is defined.
+This design choice helps to prevent global variable scope cluttering, ensuring that the variables
+are contained within the context of the specific element they are associated with.
+
+### Dynamic variable handling
+
+Execution Listeners enable the dynamic fetching and calculation of variables for the next steps in the process,
+based on the current data. This functionality is particularly useful for elements such as gateways and events,
+allowing for more informed decision-making and precise control over process flow.
```
:wrench: The usage of ELs and its technical base is most relevant for the user. We should move his part to the top.

:wrench: We should improve the wording to make it easier for users to understand.

saig0 reviewed (changes requested):
@ce-dmelnych thank you for writing the docs. It's a good start. :+1: 

However, I see the potential to make it more effective for users by rephrasing and restructuring some parts. I will schedule a meeting with you to discuss the improvements.

saig0 commented:
@ce-dmelnych I pushed the changes from our review meeting today. :rocket:

christinaausley reviewed (approved):
This is so clean and well documented! Great work 🚀 

Should we also add this to the [glossary](https://docs.camunda.io/docs/reference/glossary/)?

ce-dmelnych commented:
Thank you both for your valuable reviews and the improvements you committed to the documentation! 🙌

@saig0 , I appreciate the meeting and the suggestions you provided, which have greatly enhanced the clarity and effectiveness of the content 🙇‍♂️ 

@christinaausley, thank you for your positive feedback!
I’ve just pushed another commit with some small improvements to the wording and formatting, fixed a broken link, and added the `Execution listener` term to the [glossary](https://preview.docs.camunda.cloud/pr-4017/docs/next/reference/glossary/#execution-listener). Please review it to see if it aligns with your expectations.

Thanks again for your support and collaboration! 🚀

megglos commented:
@saig0 @ce-dmelnych  do you consider it doable to get this merged by Tuesday next week? Will include a link in the release blogpost then.

saig0 commented:
@megglos yes. I'll review it on Monday.

saig0 reviewed (approved):
LGTM. :rocket:
